src/xserver/mod.rs

```diff
@@ -1,22 +1,229 @@
+        unsafe {
+            // Open the display
+        
+            // Get the window attributes
+            let mut window_attributes: xlib::XWindowAttributes = std::mem::zeroed();
+            xlib::XGetWindowAttributes(display,  self.window, &mut window_attributes);
+    
+            // Get the dimensions of the window
+            let width = window_attributes.width as u32;
+            let height = window_attributes.height as u32;
+            
+            // Create an XImage structure to hold the screenshot
+            let image = xlib::XGetImage(
+                display,
+                self.window,
+                0,
+                0,
+                width,
+                height,
+                xlib::XAllPlanes(),
+                xlib::ZPixmap,
+            );
+    
+            // Process the image data
+            let mut buffer: Vec<u8> = Vec::with_capacity((width * height * 4) as usize);
+    
+            for y in 0..height {
+                for x in 0..width {
+                    let pixel = xlib::XGetPixel(image, x as i32, y as i32);
+                    buffer.push(((pixel >> 16) & 0xFF) as u8); // Red
+                    buffer.push(((pixel >> 8) & 0xFF) as u8);  // Green
+                    buffer.push((pixel & 0xFF) as u8);         // Blue
+                }
+            }
+    
+            // Create an ImageBuffer from the pixel data
+            let image_buffer = ImageBuffer::<Rgb<u8>, _>::from_raw(width, height, buffer).unwrap();
+            
+            // Clean up
+            xlib::XDestroyImage(image);
+            return Ok(image_buffer);
+        }
+    }
+    fn _run(&self, send: SyncSender<ToBrain>) -> eyre::Result<()> {
+        let display =unsafe { xlib::XOpenDisplay(ptr::null())};
+        loop {
+            send.send(ToBrain::NextFrame(self.get_image(display)?))?;
+        }
+    }
+}
+
+impl XController {
+    pub fn new(window_name: &str) -> eyre::Result<Self>{
+        Ok(Self {window: get_window_id_by_title(window_name)?})
+    }
+
+    
+    pub fn move_mouse_to_coordinate(&self, display: *mut _XDisplay, x: i32, y: i32) {
+        unsafe {   
+            // Move the mouse to the specified coordinates
+            xlib::XWarpPointer(display, 0, self.window, 0, 0, 0, 0, x, y);
+            xlib::XFlush(display);
+        }
+    }
+    pub fn left_click(&self, display: *mut _XDisplay, x: i32, y: i32) {
+        unsafe {
+            {
+            // Create a button press event
+            let mut button_event: xlib::XButtonEvent = std::mem::zeroed();
+            button_event.type_ = xlib::ButtonPress;
+            button_event.display = display;
+            button_event.window = self.window;
+            button_event.subwindow = 0;
+            button_event.time = 0;
+            button_event.x = x;
+            button_event.y = y;
+            button_event.same_screen = xlib::True;
+            button_event.button = 1; // 1 corresponds to the left mouse button
```

Comment:
This should be ```XButtonEvent { type_: ..., display: ..., /* other fields */, .. Default::default() }```. There is no need to use unsafe code for POD struct initialization. This also applies to other structs instantiated in this method and `cast_hook`.